### PR TITLE
add button

### PR DESCRIPTION
**Pull request by piensaenpixel**
@javierarce i suppose you need add now some logic, all yours (when you can)

image

### CODE REVIEW

**Comment by javierarce** [comment]
@xavijam: can you review my javascript changes, please? I added the logic to create widgets, layers and elements (this two last items will be available in the future)

**Comment by javierarce** [comment]
Ping?

File: lib/assets/javascripts/cartodb3/components/tab-pane/create-text-labels-tab-pane.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ module.exports = function (paneItems, options) {
     });
 
     return {
+      kind: paneItem.kind,
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/tab-pane/create-text-labels-tab-pane.js]: Shouldn't it be 'context' more than `kind`? Tab panes only knows about content or context, not sure :disappointed:.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/tab-pane/create-text-labels-tab-pane.js]: or name!

File: lib/assets/javascripts/cartodb3/editor/editor-view.js [hunk 2]
```diff
@@ -24,21 +28,47 @@ module.exports = cdb.core.View.extend({
     this._mapStackLayoutModel = opts.mapStackLayoutModel;
   },
 
+  _addItem: function () {
+    var selectedTab = this._mapTabPaneView.collection.getSelected();
+
+    switch (selectedTab.get('kind')) {
+      case 'widgets':
+        this._editorWidgetsView.addWidget();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: We should be able to open the dialog in this view, shouldn't we?

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: Not sure if this is the best way, modals master @viddo ?

Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: > We should be able to open the dialog in this view, shouldn't we?

Yes, and the add-side-effect should not rely on _content views'_ state at all, the `addWidget` logic should be moved to this view instead.

File: lib/assets/javascripts/cartodb3/editor/editor-view.js [hunk 3]
```diff
@@ -24,21 +28,47 @@ module.exports = cdb.core.View.extend({
...
+      case 'elements':
+        // TODO: trigger element creation
+        break;
+    }
+  },
+
   render: function () {
     var self = this;
 
+    this._editorWidgetsView = new EditorWidgetsView({
```
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: Related to my [prev comment](https://github.com/CartoDB/cartodb/pull/6972/files#r58026109), these views should not need to be statically defined here.

Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: :ok_hand:

File: lib/assets/javascripts/cartodb3/editor/editor-view.js (hunk 2 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: :ok_hand:

File: lib/assets/javascripts/cartodb3/editor/tab-pane.tpl [hunk 4]
```diff
@@ -1,2 +1,9 @@
-<ul class="CDB-NavMenu-Inner CDB-Text is-semibold CDB-Size-medium js-menu"></ul>
+<nav class="CDB-NavMenu">
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/tab-pane.tpl]: Add button shouldn't be added by default in the template.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/tab-pane.tpl]: Add button shouldn't be added by default in the template.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/tab-pane.tpl]: I'd try to keep this file as easy as possible. If we need a different template for another purpose, let's create a new one and pass it as a new argument.

**Comment by viddo** [comment]
Now that the add-button is added to the tab-pane context, shouldn't the add-button in the widgets listing be removed? AFAIK that was just a temporary thing until this was to be done.

**Comment by javierarce** [comment]
You are right, @viddo, I'll get rid of those buttons.

File: lib/assets/javascripts/cartodb3/editor/tab-pane.tpl (hunk 4 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/tab-pane.tpl]: But this one is just that, isn't it? It's a copy of `lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane.tpl` with the button and the classes for this particular view.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/tab-pane.tpl]: Ahhh, sorry, I saw the name and I mixed up both. Could we rename it? :sweat_smile:

**Comment by xavijam** [comment]
BTW, I've realized that button should appear only when there is any element already added to the current context, right @piensaenpixel?

**Comment by piensaenpixel** [comment]
Yes @xavijam

**Comment by viddo** [comment]
What I was referring to was the [temporary add-widget-button](https://github.com/CartoDB/cartodb/blob/add-button/lib/assets/javascripts/cartodb3/editor/widgets/widgets-view.js#L71-L81), not only the layer one

**Comment by javierarce** [comment]
@viddo: I'm afraid that method is called from the other button
 
screen shot 2016-03-31 at 17 22 55

**Comment by javierarce** [comment]
items

File: lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane-view.js [hunk 5]
```diff
@@ -42,6 +51,7 @@ module.exports = cdb.core.View.extend({
         this.removeView(this._selectedView);
       }
 
+      this.trigger('changeTab', model, this);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane-view.js]: Why do we need this?

File: lib/assets/javascripts/cartodb3/editor/editor-view.js [hunk 6]
```diff
@@ -100,9 +111,82 @@ module.exports = cdb.core.View.extend({
...
+  _showAddButton: function () {
+    this.$('.js-add').removeClass('is-hidden');
+  },
+
+  _updateAddButtonState: function () {
+    this._hideAddButton();
+
+    switch (this._mapTabPaneView.getSelectedTabPaneName()) {
+      case 'widgets':
+        if (this._widgetDefinitionsCollection.length) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: If it is a collection, this .lenght thing shouldn't work. We should use .size() instead.

File: lib/assets/javascripts/cartodb3/editor/editor-view.js [hunk 7]
```diff
@@ -100,9 +111,82 @@ module.exports = cdb.core.View.extend({
...
+          this._showAddButton();
+        }
+        break;
+      case 'layers':
+        if (this._layerDefinitionsCollection.length) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/editor-view.js]: ^^^

**Comment by xavijam** [comment]
LGTM :ship: :it: after reviewing my last comment about .length :palm_tree:

File: lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane-view.js (hunk 5 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane-view.js]: To know that the user changed the tab and toggle the visibility of the add button.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/tab-pane/tab-pane-view.js]: we could know it when tabs collection changed the selected item.

**Comment by javierarce** [comment]
Merging this since the frontend builds are :green_apple:

**Comment by xavijam** [comment]
:+1:

**Pull request merged**